### PR TITLE
Add more content on configurations

### PR DESCRIPTION
**Pull request by bmuschko**
### Context

See https://github.com/gradle/dotorg-docs/issues/179

The terms "configuration" and "configurations" is among the top 20 user-based search terms. This change adds more content on certain aspects around configurations.

- Cross-reference user guide from DSL guide/Javadocs. (Those links may easily break in the future. There's no automation around verifying the existence or pass in a specific version of the user guide.)
- More references to API from user guide
- Explain the concept of configuration inheritance
- More cross-references to the basic "configuration" information
- Additional use cases in "Working with Dependencies" chapter. More complex code samples use the concrete types of the API to give a better hint to the referenced classes.

### CODE REVIEW

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/Configuration.java [hunk 1]
```diff
@@ -295,16 +297,24 @@ public String determineName(Configuration c) {
     /**
      * Gets the set of dependencies directly contained in this configuration
      * (ignoring superconfigurations).
+     * <p>
+     * This method does not resolve the configuration. Therefore, the return values does not include
```
Comment (oehme) [inline subprojects/core-api/src/main/java/org/gradle/api/artifacts/Configuration.java]: ❌ I'd rephrase this and use the term "declared" dependencies here to make the distinction from "resolved" more obvious. The objects here are completely different from the ones you get in the resolution result. Depending on how the Configuration is resolved, they can end up pointing to something completely different, e.g. because of substitutions.

File: subprojects/docs/src/docs/userguide/declaringDependencies.adoc [hunk 2]
```diff
@@ -179,6 +183,32 @@ Let's say you wanted to declare a dependency on the link:https://tomcat.apache.o
...
+
+A configuration can extend other configurations to form an inheritance hierarchy similar to class inheritance in object-oriented programming. Configurations with a super configuration have access to the whole set of dependencies assigned to the hierarchy.
+
+This concept is used by Gradle core plugins like the <<sec:java_plugin_and_dependency_management,Java plugin>>. For example the `testCompile` configuration extends the `compile` configuration. Compiling tests requires the dependencies of the source code under test to be available on the classpath on top of the dependencies needed to compile the test code. You could easy imagine a Java project that uses the Guava dependency in the production source code and JUnit to write and execute test code.
+
++++++
+<figure xmlns:xi="http://www.w3.org/2001/XInclude">
+    <title>Configuration inheritance provided by the Java plugin</title>
+    <imageobject>
+        <imagedata fileref="img/dependency-management-configuration-inheritance.png" width="140mm" />
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/declaringDependencies.adoc]: ❌ Should use implementation/testImplementation, not the deprecated compile/testCompile

File: subprojects/docs/src/docs/userguide/workingWithDependencies.adoc [hunk 3]
```diff
@@ -46,6 +46,18 @@ You can iterate over the complete set of artifacts resolved for a module with th
 Iterating over the artifacts of a module automatically resolves the configuration. A resolved configuration becomes immutable and cannot add or remove dependencies. If needed you can copy a configuration for further modification via api:org.gradle.api.artifacts.Configuration#copy()[].
 ====
 
+=== Navigating the dependency graph
+
+As a plugin developer, you might be faced with the challenge of navigating the full graph of dependencies assigned to a configuration e.g. for turning the dependency graph into a visualization. You can access the full graph of dependencies for a configuration with the help of the api:org.gradle.api.artifacts.result.ResolutionResult[].
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/workingWithDependencies.adoc]: ⭕️ "faced with the challenge" sounds odd to me - the API makes this super simple :)

File: subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/iterateDependencies/build.gradle [hunk 4]
```diff
@@ -12,12 +12,14 @@ dependencies {
 }
 
 //START SNIPPET iteration-task
-task iterateScmDependencies {
+task iterateDeclaredDependencies {
     dependsOn configurations.scm
```
Comment (oehme) [inline subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/iterateDependencies/build.gradle]: ❌ This dependsOn is counterproductive. Since the task only iterates over the declared dependencies, it doesn't need the configuration to be resolved before. This will just make this task run slower than it needs to.

File: subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/walkGraph/build.gradle [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+//START SNIPPET walk-task
+task walkDependencyGraph {
+    dependsOn configurations.scm
+
+    doLast {
+        ResolutionResult resolutionResult = configurations.scm.incoming.resolutionResult
+
+        resolutionResult.allDependencies {
```
Comment (oehme) [inline subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/walkGraph/build.gradle]: ❌  This uses the graph as a flat list, which is better served by the `incoming.artifacts` API, which needs less memory. You should only use the graph if you actually care about the shape of the graph. E.g. you could print the components with increasing levels of indentation depending on where in the graph they were found.

File: subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/accessMetadataArtifact/build.gradle [hunk 6]
```diff
@@ -13,16 +13,17 @@ task printGuavaMetadata {
...
 
     doLast {
-        def resolutionResult = configurations.compileClasspath.incoming.resolutionResult
-        def componentIds = resolutionResult.allDependencies.collect { it.selected.id }
+        ResolutionResult resolutionResult = configurations.compileClasspath.incoming.resolutionResult
+        List<ComponentIdentifier> componentIds = resolutionResult.allDependencies.collect { it.selected.id }
 
-        def result = dependencies.createArtifactResolutionQuery().forComponents(componentIds)
+        ArtifactResolutionQuery query = dependencies.createArtifactResolutionQuery()
+            .forComponents(componentIds)
```
Comment (oehme) [inline subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/accessMetadataArtifact/build.gradle]: ❌ This is overcomplicated and inefficient if you really only want the pom for guava, you can just pass the String `com.google:guava:18.0` here. No need to first get the dependency graph (which is very expensive).

File: subprojects/docs/src/docs/userguide/workingWithDependencies.adoc [hunk 7]
```diff
@@ -15,13 +15,13 @@
 [[working_with_dependencies]]
 == Working with Dependencies
 
-Gradle provides an extensive API for navigating, inspecting and post-processing post-processing metadata and artifacts of resolved dependencies. With this API at hand, you can implement more complex requirements to your build requiring a "runtime view" of dependencies.
+Gradle provides an extensive API for navigating, inspecting and post-processing metadata and artifacts of resolved dependencies. With this API at hand, you can implement more complex requirements to your build requiring a "runtime view" of dependencies.
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/workingWithDependencies.adoc]: ❌ you can implement more complex requirements to your build requiring a "runtime view" of dependencies.

That's a mouthful, can we say this simpler?

File: subprojects/docs/src/docs/userguide/declaringDependencies.adoc [hunk 8]
```diff
@@ -179,6 +183,32 @@ Let's say you wanted to declare a dependency on the link:https://tomcat.apache.o
 
 A project's configurations are managed by a `configurations` object. Configurations have a name and can extend each other. To learn more about this API have a look at api:org.gradle.api.artifacts.ConfigurationContainer[].
 
+[[sub:inheriting_dependencies_from_other_configurations]]
+==== Inheriting dependencies from other configurations
+
+A configuration can extend other configurations to form an inheritance hierarchy similar to class inheritance in object-oriented programming. Configurations with a super configuration have access to the whole set of dependencies assigned to the hierarchy.
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/declaringDependencies.adoc]: ⭕️ I'd leave out the OO reference, it's not a strong analogy.

File: subprojects/docs/src/docs/userguide/declaringDependencies.adoc [hunk 9]
```diff
@@ -179,6 +183,32 @@ Let's say you wanted to declare a dependency on the link:https://tomcat.apache.o
...
+
+This concept is used by Gradle core plugins like the <<sec:java_plugin_and_dependency_management,Java plugin>>. For example the `testCompile` configuration extends the `compile` configuration. Compiling tests requires the dependencies of the source code under test to be available on the classpath on top of the dependencies needed to compile the test code. You could easy imagine a Java project that uses the Guava dependency in the production source code and JUnit to write and execute test code.
```
[the remark this answers is not in the document]
Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/declaringDependencies.adoc]: Sure, will change that. I still don't understand why using `compile` is a bad thing when building a Java application. My argument would be: "if it's a bad thing then we why is it not deprecated?". Can somebody write this down in a couple of concrete sentences?

File: subprojects/docs/src/docs/userguide/workingWithDependencies.adoc [hunk 10]
```diff
@@ -15,13 +15,13 @@
...
+=== Iterating over dependencies assigned to a configuration
 
-You can iterate over complete set of dependencies assigned to a configuration with the help of the method api:org.gradle.api.artifacts.Configuration#getDependencies()[].
+Sometimes you'll want to implement logic based on the declared dependencies of a project e.g. to inspect the declared dependencies in build script from a Gradle plugin. You can iterate over set of dependencies assigned to a configuration with the help of the method api:org.gradle.api.artifacts.Configuration#getDependencies()[]. Alternatively, you can also use api:org.gradle.api.artifacts.Configuration#getAllDependencies()[] to include the declared dependencies of <<sub:inheriting_dependencies_from_other_configurations,superconfigurations>>. The returned dependencies do not resolve the configuration and therefore do not include transitive dependencies.
```
[the remark this answers is not in the document]
Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/workingWithDependencies.adoc]: done

File: subprojects/docs/src/docs/userguide/declaringDependencies.adoc (hunk 9 above)
[the remark this answers is not in the document]
Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/declaringDependencies.adoc]: done

File: subprojects/docs/src/docs/userguide/declaringDependencies.adoc [hunk 11]
```diff
@@ -165,6 +167,8 @@ Many Gradle plugins add pre-defined configurations to your project. The Java plu
 </figure>
 +++++
 
+For more examples on the usage of configurations to navigate, inspect and post-processing metadata and artifacts of assigned dependencies, see <<working_with_dependencies>>.
```
[the remark this answers is not in the document]
Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/declaringDependencies.adoc]: done

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/Configuration.java [hunk 12]
```diff
@@ -295,16 +297,24 @@ public String determineName(Configuration c) {
...
      */
     DependencySet getDependencies();
 
     /**
-     * <p>Gets the complete set of dependencies including those contributed by
-     * superconfigurations.</p>
+     * Gets the complete set of dependencies including those contributed by
+     * superconfigurations.
+     * <p>
+     * This method does not resolve the configuration. Therefore, the return values does not include
```
[the remark this answers is not in the document]
Comment (bmuschko) [reply subprojects/core-api/src/main/java/org/gradle/api/artifacts/Configuration.java]: done

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/Configuration.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (bmuschko) [reply subprojects/core-api/src/main/java/org/gradle/api/artifacts/Configuration.java]: done

File: subprojects/docs/src/docs/userguide/declaringDependencies.adoc (hunk 2 above)
Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/declaringDependencies.adoc]: done. We'll need to do something about the documentation on that front. It's absolutely unclear to a user that these configurations are deprecated or "somewhat, lightly deprecated". The documentation should say "Don't use these because of x. Use y instead.". Are there still _any_ uses cases where you'd want to use `compile` or `testCompile`?

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/Configuration.java (hunk 1 above)
Comment (bmuschko) [reply subprojects/core-api/src/main/java/org/gradle/api/artifacts/Configuration.java]: done

File: subprojects/docs/src/docs/userguide/declaringDependencies.adoc (hunk 8 above)
Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/declaringDependencies.adoc]: done

File: subprojects/docs/src/docs/userguide/workingWithDependencies.adoc (hunk 7 above)
Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/workingWithDependencies.adoc]: I think it's somewhat redundant. I am going to remove it. Reading the sentence also make me notice that it doesn't flow that well.

File: subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/iterateDependencies/build.gradle (hunk 4 above)
Comment (bmuschko) [reply subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/iterateDependencies/build.gradle]: I made this change based on a comment from a previous review: https://github.com/gradle/gradle/pull/4124#discussion_r163356699. It's somewhat not relevant to the example. However, I know that a lot of people just copy/paste whatever snippet they find on the web and then complain that it doesn't work properly in other contexts.

File: subprojects/docs/src/docs/userguide/workingWithDependencies.adoc (hunk 3 above)
Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/workingWithDependencies.adoc]: done

File: subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/accessMetadataArtifact/build.gradle (hunk 6 above)
Comment (bmuschko) [reply subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/accessMetadataArtifact/build.gradle]: Can you clarify? I don't understand what method you are referring to when you say "pass in the String". The method `fromComponents` only takes one or many `ComponentIdentifier`s. If we want to use a `ComponentIdentifier` then we'd have to create a new one via an internal API.

Comment (oehme) [reply subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/accessMetadataArtifact/build.gradle]: fromComponents was changed to take Strings. Unless we removed that again @jjohannes?

File: subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/iterateDependencies/build.gradle (hunk 4 above)
Comment (oehme) [reply subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/iterateDependencies/build.gradle]: That was for iterating the artifacts. But herr you just want declared dependencies. Depending in the configuration would download the artifacts even though your task doesn't need them at all. Maybe replace this with an example that isn't a task to make clear that this is safe to call at any time?

File: subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/accessMetadataArtifact/build.gradle (hunk 6 above)
Comment (bmuschko) [reply subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/accessMetadataArtifact/build.gradle]: I don't see that method in https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/query/ArtifactResolutionQuery.html or on `master`.

File: subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/walkGraph/build.gradle (hunk 5 above)
Comment (bmuschko) [reply subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/walkGraph/build.gradle]: > This uses the graph as a flat list, which may be better served by the incoming.artifacts API in case you also want the artifacts.

The intend was not to render artifacts here.

> I think this could be a bit clearer if the example actually showed a graph traversal, e.g. printing components at different indentation levels based on a breadth-first search.

I'll work on that.

Comment (bmuschko) [reply subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/walkGraph/build.gradle]: done

File: subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/iterateDependencies/build.gradle (hunk 4 above)
Comment (bmuschko) [reply subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/iterateDependencies/build.gradle]: Removed. I added a sentence which clearly states that it is safe to call this during configuration time.

**Comment by bmuschko** [comment]
> I also like the cross-referencing from the javadocs. Can we create a follow up issue to establish some automating link testing in the future?

https://github.com/gradle/dotorg-docs/issues/186

File: subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/accessMetadataArtifact/build.gradle (hunk 6 above)
Comment (jjohannes) [reply subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/accessMetadataArtifact/build.gradle]: I don't know anything about this. It does not look like any change made it to master. Do you remember in which context we wanted to change this @oehme?

File: subprojects/docs/src/docs/userguide/declaringDependencies.adoc (hunk 2 above)
Comment (jjohannes) [reply subprojects/docs/src/docs/userguide/declaringDependencies.adoc]: I think we should continue the discussion here and address this as part of the "Update Java (Library) plugin chapter" story: https://github.com/gradle/dotorg-docs/issues/42

File: subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/accessMetadataArtifact/build.gradle (hunk 6 above)
Comment (bmuschko) [reply subprojects/docs/src/samples/userguide/dependencyManagement/workingWithDependencies/accessMetadataArtifact/build.gradle]: The method is called `forModule`: https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/query/ArtifactResolutionQuery.html#forModule-java.lang.String-java.lang.String-java.lang.String-

File: subprojects/docs/src/docs/userguide/workingWithDependencies.adoc [hunk 13]
```diff
@@ -46,13 +46,25 @@ You can iterate over the complete set of artifacts resolved for a module with th
 Iterating over the artifacts of a module automatically resolves the configuration. A resolved configuration becomes immutable and cannot add or remove dependencies. If needed you can copy a configuration for further modification via api:org.gradle.api.artifacts.Configuration#copy()[].
 ====
 
+=== Navigating the dependency graph
```
Comment (eriwen) [inline subprojects/docs/src/docs/userguide/workingWithDependencies.adoc]: Seems like we should consider showing the dependency tree of build scans here.

Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/workingWithDependencies.adoc]: I am not quite sure how a build scan would be relevant for explaining this API. We actually do want to walk the graph programmatically. The example I use here is that you are able to create a _different_ visualization than provided by a build scan e.g. an actual DAG. I think the story would be different if a build scan would provide such a visualization (especially if you could zoom in/out etc.) but at the moment it doesn't. What were your thoughts here?
